From b3b942a04b0f943ea11d63e9269fcd49356174c9 Mon Sep 17 00:00:00 2001
From: Manikanta Mylavarapu <quic_mmanikan@quicinc.com>
Date: Mon, 26 Aug 2024 18:14:47 +0530
Subject: [PATCH] cpufreq: qcom-nvmem: enable cpufreq for ipq54xx

Unlike previous soc's, ipq54xx speedbin can be directly
compared with hardware configuration (PLL Lvalue). Hence
assign speedbin value directly to version.

Change-Id: I02dc37953c8d8df074068772f4d92fc3218bc371
Signed-off-by: Manikanta Mylavarapu <quic_mmanikan@quicinc.com>
---
 drivers/cpufreq/cpufreq-dt-platdev.c | 1 +
 drivers/cpufreq/qcom-cpufreq-nvmem.c | 8 +++++++-
 2 files changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/cpufreq/cpufreq-dt-platdev.c b/drivers/cpufreq/cpufreq-dt-platdev.c
index 1a25d962ce2e..d34c7d7074fa 100644
--- a/drivers/cpufreq/cpufreq-dt-platdev.c
+++ b/drivers/cpufreq/cpufreq-dt-platdev.c
@@ -178,6 +178,7 @@ static const struct of_device_id blocklist[] __initconst = {
 	{ .compatible = "ti,am62a7", },
 
 	{ .compatible = "qcom,ipq5332", },
+	{ .compatible = "qcom,ipq5424", },
 	{ .compatible = "qcom,ipq8064", },
 	{ .compatible = "qcom,ipq9574", },
 	{ .compatible = "qcom,apq8064", },
diff --git a/drivers/cpufreq/qcom-cpufreq-nvmem.c b/drivers/cpufreq/qcom-cpufreq-nvmem.c
index e1a3774d5ed0..a8a6ce33f2d1 100644
--- a/drivers/cpufreq/qcom-cpufreq-nvmem.c
+++ b/drivers/cpufreq/qcom-cpufreq-nvmem.c
@@ -30,6 +30,9 @@
 
 #include <dt-bindings/arm/qcom,ids.h>
 
+#define IPQ5424_BLNK_PART	0xFF
+#define IPQ5424_TURBO_FREQ_LVAL	0x4B
+
 struct qcom_cpufreq_drv;
 
 struct qcom_cpufreq_match_data {
@@ -163,7 +166,10 @@ static int qcom_cpufreq_kryo_name_version(struct device *cpu_dev,
 		break;
 	case QCOM_ID_IPQ5424:
 	case QCOM_ID_IPQ5404:
-		drv->versions = 1 << (unsigned int)!!(*speedbin);
+		drv->versions = (unsigned int)(*speedbin);
+		/* Configure blank parts with turbo frequency */
+		if (drv->versions == IPQ5424_BLNK_PART)
+			drv->versions = IPQ5424_TURBO_FREQ_LVAL;
 		break;
 	case QCOM_ID_MSM8996SG:
 	case QCOM_ID_APQ8096SG:
-- 
2.34.1

